platforms/chibios/eeprom_stm32_l4.c

```diff
@@ -0,0 +1,546 @@
+    return DataByte;
+}
+
+uint16_t EEPROM_ReadDataWord(uint16_t Address) {
+    uint16_t DataWord = 0xFFFF;
+
+    if (Address < FEE_DENSITY_BYTES - 1) {
+        /* Check word alignment */
+        if (Address % 2) {
+            DataWord = DataBuf[Address] | (DataBuf[Address + 1] << 8);
+        } else {
+            DataWord = *(uint16_t *)(&DataBuf[Address]);
+        }
+    }
+
+    return DataWord;
+}
+
+/*****************************************************************************
+ *  Bind to eeprom_driver.c
+ *******************************************************************************/
+void eeprom_driver_init(void) { EEPROM_Init(); }
+
+void eeprom_driver_erase(void) { EEPROM_Erase(); }
+
+void eeprom_read_block(void *buf, const void *addr, size_t len) {
+    const uint8_t *src  = (const uint8_t *)addr;
+    uint8_t *      dest = (uint8_t *)buf;
+
+    /* Check word alignment */
+    if (len && (uint32_t)src % 2) {
+        /* Read the unaligned first byte */
+        *dest++ = EEPROM_ReadDataByte((const uintptr_t)((uint16_t *)src));
+        --len;
+    }
+
+    uint16_t value;
+    bool     aligned = ((uint32_t)dest % 2 == 0);
+    while (len > 1) {
+        value = EEPROM_ReadDataWord((const uintptr_t)((uint16_t *)src));
+        if (aligned) {
+            *(uint16_t *)dest = value;
+            dest += 2;
+        } else {
+            *dest++ = value;
+            *dest++ = value >> 8;
+        }
+        src += 2;
+        len -= 2;
+    }
+    if (len) {
+        *dest = EEPROM_ReadDataByte((const uintptr_t)src);
+    }
+}
+
+void eeprom_write_block(const void *buf, void *addr, size_t len) {
+    uint8_t *      dest = (uint8_t *)addr;
+    const uint8_t *src  = (const uint8_t *)buf;
+    uint8_t        write_len;
+
+    while (len > 0) {
+        /* Check and try to write double word fisrt */
+        if ((uintptr_t)dest % 4 == 0 && len >= 4) {
+            uint32_t dwvalue;
+            bool     dwaligned = ((uint32_t)src % 4 == 0);
+
+            if (dwaligned) {
+                dwvalue = *(uint32_t *)src;
+            } else {
+                dwvalue = *(uint8_t *)src | (*(uint8_t *)(src + 1) << 8) | (*(uint8_t *)(src + 2) << 16) | (*(uint8_t *)(src + 3) << 24);
+            }
+            EEPROM_WriteDataDWord((uintptr_t)((uint16_t *)dest), dwvalue);
+            write_len = 4;
+        }
+        /* Check and try to write word */
+        else if ((uintptr_t)dest % 2 == 0 && len >= 2) {
+            uint16_t wvalue;
+            bool     waligned = ((uintptr_t)src % 2 == 0);
+
+            if (waligned) {
+                wvalue = *(uint16_t *)src;
+            } else {
+                wvalue = *(uint8_t *)src | (*(uint8_t *)(src + 1) << 8);
+            }
+            EEPROM_WriteDataWord((uintptr_t)((uint16_t *)dest), wvalue);
+            write_len = 2;
+        } else {
+            /* Write the unaligned or single byte */
+            EEPROM_WriteDataByte((uintptr_t)dest++, *src++);
+            write_len = 1;
```

Comment:
there's a bug here where you end up incrementing dest and src twice, first the inline post-increment and then when you do `dest/src += write_len` at the end of the loop